### ISSUE TITLE
[Bug?]: `ENOENT` in Redwood contributing workflow using Yarn

### ISSUE THREAD

**Issue by jtoar** [yarnpkg/berry#4429]
### Self-service

- [X] I'd be willing to implement a fix

### Describe the bug

Hey everyone. Opening this because I'm seeing a `YN0001` error code, but I'm not entirely convinced this is a bug in Yarn yet.

I work on [Redwood](https://github.com/redwoodjs/redwood); there's an issue in our contributing workflow that's slowing down development. I'll write it up as best I can below, but here's a link to a video of the bug that I recommend watching if you can: https://s.tape.sh/9hTynWoS.

Both the Redwood framework and Redwood projects use Yarn 3, but they use the `"node-modules"` `nodeLinker` because we don't follow the [new rules](https://yarnpkg.com/advanced/rulebook) yet.

Right now, to test new features in the framework on projects, we copy the framework's packages' `dist` to projects `node_modules`. So we're modifying `node_modules` from under Yarn. (Note that we're only doing this because we couldn't seem to get `yarn link` to work, maybe because we don't follow the new rules yet?)

We made our own contributing tool to do this, and while it mostly works quite well, if you try to `yarn install` _after_ `node_modules` has been modified by our contributing tool, Yarn throws a `YN0001` error code during the link step:

```
➤ YN0000: ┌ Link step
➤ YN0001: │ Error: ENOENT: no such file or directory, lstat '/Users/dom/prjcts/redwood_project/redwood_project/node_modules/@redwoodjs/core/node_modules/semver'
➤ YN0000: └ Completed in 0s 636ms
```

Is this actually a bug in yarn? Or are we breaking the contract by modifying `node_modules` and expecting `yarn install`s to still work after?

Lastly, thanks again for when we migrated from yarn 1 to yarn 3—everyone was amazingly helpful!

### To reproduce

I'll try to get this working in Sherlock, but in the meantime the video I linked to above shows the exact steps I took: https://s.tape.sh/9hTynWoS. And the issue should be reproducible locally by following the steps I took:

```bash
# Clone the Redwood framework, install deps and build
git clone https://github.com/redwoodjs/redwood.git
cd redwood
yarn
yarn build

# Make a redwood project
cd ~
yarn create redwood-app redwood-project

# Contributing workflow; `RWFW_PATH` is the path to your local copy of the framework
cd redwood-project
RWFW_PATH=~/redwood yarn rwfw project:deps && yarn
RWFW_PATH=~/redwood yarn rwfw project:copy

# Now installing again should throw the error
yarn
...
➤ YN0000: ┌ Link step
➤ YN0001: │ Error: ENOENT: no such file or directory, lstat '/Users/dom/prjcts/redwood_project/redwood_project/node_modules/@redwoodjs/core/node_modules/semver'
➤ YN0000: └ Completed in 0s 636ms
```

### Environment

```shell
System:
    OS: macOS 12.1
    CPU: (8) arm64 Apple M1
  Binaries:
    Node: 16.15.0 - /private/var/folders/dt/yks4v5m53k114qxgz6jh4pgw0000gn/T/xfs-e9f6d41d/node
    Yarn: 3.2.0 - /private/var/folders/dt/yks4v5m53k114qxgz6jh4pgw0000gn/T/xfs-e9f6d41d/yarn
    npm: 8.5.5 - ~/.nvm/versions/node/v16.15.0/bin/npm
  npmPackages:
    jest: 27.5.1 => 27.5.1
```

**Comment by arcanis** [comment]
Modifying the node_modules content isn't supported, indeed. Changes should be made using preferably `portal:` (to follow dependencies), `link:` (to create folder aliases), or `file:` (to copy a package in the nm rather than symlinking it; kinda discouraged).

What were the problems you hit with `yarn link`?

**Comment by merceyz** [comment]
> Is this actually a bug in yarn?

@jtoar It was, @larixer fixed it in https://github.com/yarnpkg/berry/pull/3467 which was first released in 4.0.0-rc.1.

> Modifying the node_modules content isn't supported, indeed

@arcanis That depends on the modification, with the linked PR `rm -r ./node_modules/lodash` will cause Yarn to recreate the folder but not `rm ./node_modules/lodash/index.js`

**Comment by jtoar** [comment]
@merceyz confirming the error is gone on yarn 4!

@arcanis with `yarn link`, I hit this immediately: 

```bash
$ pwd
redwood_project

$ yarn link ../redwood_framework --all
...
➤ YN0000: ┌ Link step
➤ YN0001: │ Error: Assertion failed: Writing attempt prevented to /Users/dom/prjcts/redwood/redwood_framework/packages/core/node_modules/@redwoodjs/cli which is outside project root: /Users/dom/prjcts/redwood/redwood_project
    at fb (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:712:1725)
    at nce (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:712:2298)
    at T_e (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:712:6802)
    at rce.finalizeInstall (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:697:26714)
    at async ze.linkEverything (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:441:14959)
    at async /Users/dom/.node/corepack/yarn/3.2.0/yarn.js:444:4324
    at async Je.startSectionPromise (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:413:3303)
    at async ze.install (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:444:4100)
    at async /Users/dom/.node/corepack/yarn/3.2.0/yarn.js:529:1744
    at async Function.start (/Users/dom/.node/corepack/yarn/3.2.0/yarn.js:413:2373)
➤ YN0000: └ Completed in 0s 719ms
➤ YN0000: Failed with errors in 7s 94ms
```

It looks like yarn is trying to write nested `node_modules` to `redwood_framework` instead of `redwood_project`? Just to provide more info, `@redwoodjs/cli` is a dependency of `@redwoodjs/core`:

```bash
$ pwd
redwood_project

$ yarn why @redwoodjs/cli
└─ @redwoodjs/core@portal:/Users/dom/prjcts/redwood/redwood_framework/packages/core::locator=root-workspace-0b6124%40workspace%3A.
   └─ @redwoodjs/cli@portal:/Users/dom/prjcts/redwood/redwood_framework/packages/cli::locator=root-workspace-0b6124%40workspace%3A. (via portal:/Users/dom/prjcts/redwood/redwood_framework/packages/cli::locator=root-workspace-0b6124%40workspace%3A.)
```

All the resolutions added by `yarn link` to `redwood_project`'s `package.json` look right. But maybe the lock file isn't being updated—should `@redwoodjs/cli` use the portal protocol below?

<img width="1522" alt="image" src="https://user-images.githubusercontent.com/32992335/166681437-a83ad54e-c17c-4621-89fa-529cc3811207.png">

Just to provide more context, here's the directory structure:

```
redwood/
├── redwood_project/
│  ├── api/
│  └── web/
└── redwood_framework/
   └── packages/
      ├── web/
      ├── testing/
      ├── telemetry/
      ├── structure/
      ├── router/
      ├── record/
      ├── prerender/
      ├── internal/
      ├── graphql-server/
      ├── forms/
      ├── create-redwood-app/
      ├── core/
      ├── codemods/
      ├── cli/
      ├── auth/
      ├── api/
      ├── api-server/
      └── eslint-config/
```

Happy to provide any other info, and contribute back to Yarn if this ends up being something I could fix!

**Comment by larixer** [comment]
@jtoar The issue with `yarn link` and interdepndent workspaces was also fixed in upcoming 4.x series:
https://github.com/yarnpkg/berry/blob/master/CHANGELOG.md#installs

**Comment by jtoar** [comment]
@larixer amazing, confirming that I'm no longer seeing that `yarn link` error on yarn 4! 

Now it seems like most of Redwood does work, but there's a few commands that I haven't gotten to yet. I haven't been able to do much digging yet so this is where I start to feel like it's more on us than on Yarn, but anything related to Webpack errors out like this (the command here is `yarn rw dev`, to start the dev server):

```
web | /Users/dom/prjcts/redwood/redwood/node_modules/webpack/lib/NormalModule.js:227
web | 			throw new TypeError(
web | 			^
web | 
web | TypeError: The 'compilation' argument must be an instance of Compilation
web |     at Function.getCompilationHooks (/Users/dom/prjcts/redwood/redwood/node_modules/webpack/lib/NormalModule.js:227:10)
web |     at /Users/dom/prjcts/redwood/redwood/node_modules/webpack/lib/HotModuleReplacementPlugin.js:767:18
web |     at Hook.eval [as call] (eval at create (/Users/dom/prjcts/rwprj/rwprj/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:100:1)
web |     at Hook.CALL_DELEGATE [as _call] (/Users/dom/prjcts/rwprj/rwprj/node_modules/tapable/lib/Hook.js:14:14)
web |     at Compiler.newCompilation (/Users/dom/prjcts/rwprj/rwprj/node_modules/webpack/lib/Compiler.js:1122:26)
web |     at /Users/dom/prjcts/rwprj/rwprj/node_modules/webpack/lib/Compiler.js:1166:29
web |     at Hook.eval [as callAsync] (eval at create (/Users/dom/prjcts/rwprj/rwprj/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
web |     at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/dom/prjcts/rwprj/rwprj/node_modules/tapable/lib/Hook.js:18:14)
web |     at Compiler.compile (/Users/dom/prjcts/rwprj/rwprj/node_modules/webpack/lib/Compiler.js:1161:28)
web |     at /Users/dom/prjcts/rwprj/rwprj/node_modules/webpack/lib/Watching.js:214:19
web | cd "/Users/dom/prjcts/rwprj/rwprj/web" && yarn cross-env NODE_ENV=development RWJS_WATCH_NODE_MODULES= webpack serve --config "/Users/dom/prjcts/redwood/redwood/packages/core/config/webpack.development.js"  exited with code 1
```

And a few things involving `core-js` error out like this:

```
TypeError: (0 , _bind.default) is not a function
    at Object.<anonymous> (/Users/dom/prjcts/redwood/redwood/node_modules/core-js-pure/internals/function-bind.js:24:29)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._compile (/Users/dom/prjcts/redwood/redwood/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Object.newLoader [as .js] (/Users/dom/prjcts/redwood/redwood/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/dom/prjcts/redwood/redwood/node_modules/core-js-pure/modules/es.function.bind.js:2:12)

```

Here `redwood/redwood` is the framework, and `rwprj/rwprj` is the project.

I'll debug this further, but are there any obvious gotchas to look out for? Redwood does include storybook which includes Webpack too—maybe that could be causing a conflict?

**Comment by merceyz** [comment]
The first error is due to multiple instances and/or versions of webpack.
The second error I would guess is because your babel config doesn't inject absolute paths making it use a different version of `core-js-pure` than expected.

**Comment by jtoar** [comment]
Thanks @merceyz. I know there's probably not one answer here, but in general, maybe the best way forward is for us to start migrating off the `"node-modules"` `nodeLinker` following these two guides?

- https://yarnpkg.com/getting-started/migration
- https://yarnpkg.com/advanced/rulebook

Also, happy to close this issue and discuss elsewhere if that's preferred since the original problem I brought up is solved and now I'm mostly just asking about best practices.

**Comment by merceyz** [comment]
I [see you noticed](https://github.com/redwoodjs/redwood/pull/5525) but for any future readers; the fix was backported and released in Yarn v3.2.1.

> maybe the best way forward is for us to start migrating off the "node-modules" nodeLinker following these two guides?

Maybe, though you might be able to solve the symlink issue with [`--preserve-symlinks`](https://nodejs.org/dist/latest-v16.x/docs/api/cli.html#--preserve-symlinks)

**Comment by jtoar** [comment]
> Maybe, though you might be able to solve the symlink issue with [--preserve-symlinks](https://nodejs.org/dist/latest-v16.x/docs/api/cli.html#--preserve-symlinks)

I'll give that a shot—thanks for all the help on this one!
